client-go/cache: fix missing delete event on replace

odinuge opened the pull request:
This fixes a race condition when a "short lived" object is created and the create event is still present on the queue when a relist replaces the state. Previously that would lead in the object being leaked.

The way this could happen is roughly;
```
1. new Object is added O, agent gets CREATED event for it
2. watch is terminated, and the agent runs a new list, L
3. CREATE event for O is still on the queue to be processed.
4. informer replaces the old data in store with L, and O is not in L
  - Since O is not in the store, and not in the list L, no DELETED event is queued
5. CREATE event for O is still on the queue to be processed.
6. CREATE event for O is processed
7. O is <leaked>; its present in the agent but not in k8s.
```

With this patch, on step 4. above, it would create a DELETED event ensuring that the object will be removed. We discovered these "leaks" on informers without resyncs, causing the cleanup no not work. With resync, the objects will be cleared when the next resync runs.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fix missing delete events on informer re-lists to ensure all delete events are correctly emitted and using the latest known object state, so that all event handlers and stores always reflect the actual apiserver state as best as possible
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

odinuge committed:
client-go/cache: fix missing delete event on replace

This fixes a race condition when a "short lived" object
is created and the create event is still present on the queue
when a relist replaces the state. Previously that would lead in the
object being leaked.

The way this could happen is roughly;

1. new Object is added O, agent gets CREATED event for it
2. watch is terminated, and the agent runs a new list, L
3. CREATE event for O is still on the queue to be processed.
4. informer replaces the old data in store with L, and O is not in L
  - Since O is not in the store, and not in the list L, no DELETED event
    is queued
5. CREATE event for O is still on the queue to be processed.
6. CREATE event for O is processed
7. O is <leaked>; its present in the cache but not in k8s.

With this patch, on step 4. above it would create a DELETED event
ensuring that the object will be removed.

Signed-off-by: Odin Ugedal <ougedal@palantir.com>
Signed-off-by: Odin Ugedal <odin@uged.al>

aojea commented:
/cc @wojtek-t

odinuge commented:
/test all

Triggering test to double check it doesn't break something else obvious. I'll try writing another test as well that cover the edge case in general as well, that pass with both this and the other alternative of removing it from the delta queue.

Thanks for pinging him @aojea! As I mentioned in the PR description I'll spend some more time gathering info and create an issue tomorrow morning when I get time (UK timezone). 😄

odinuge commented:
Filed https://github.com/kubernetes/kubernetes/issues/115658 so we can discuss the actual/possible issue there. 😄

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo_test.go:
```diff
@@ -121,6 +121,35 @@ func TestDeltaFIFO_replaceWithDeleteDeltaIn(t *testing.T) {
+		}),
+	})
+
+	f.Add(newObj)
```
You need to test all possible deltas -- add, update, sync (?), and delete. I think this code does the wrong thing if there's a delete already there?

odinuge replied:
> You need to test all possible deltas -- add, update, sync (?), and delete.

Yeah, smart! Will add another set of tests.

>  I think this code does the wrong thing if there's a delete already there?

If there is a delete there already, and the object should be deleted, it should be on the tail of the deltas. And if that is the case, the new "delete" event will be a noop since its deduped here I [believe](https://github.com/kubernetes/kubernetes/blob/v1.26.0/staging/src/k8s.io/client-go/tools/cache/delta_fifo.go#L396-L405).

odinuge replied:
Ahh, yee, I think you are correct. If the delete event is a normal delete event its fine, but if the object is a `
DeletedFinalStateUnknown` we have to extract the actual object. Or can we just `continue` the loop? I think for the end users both should be the same.

odinuge replied:
OOh, well, that uncover the same "issue" for the case where `f.knownObjects == nil` as well. I'll write a test to verify the behavior.

aojea commented:
/kind bug
/release-note none
/triage accepted

odinuge committed:
client-go/cache: fix missing delete event on replace without knownObjects

This fixes an issue where a relist could result in a DELETED delta
with an object wrapped in a DeletedFinalStateUnknown object; and then on
the next relist, it would wrap that object inside another
DeletedFinalStateUnknown, leaving the user with a "double" layer
of DeletedFinalStateUnknown's.

Signed-off-by: Odin Ugedal <ougedal@palantir.com>
Signed-off-by: Odin Ugedal <odin@uged.al>

odinuge committed:
client-go/cache: merge ReplaceMakesDeletionsForObjectsInQueue tests

Signed-off-by: Odin Ugedal <ougedal@palantir.com>
Signed-off-by: Odin Ugedal <odin@uged.al>

aojea commented:
can you undraft it and move it to review?

odinuge commented:
> can you undraft it and move it to review?

Ack, thanks. I'll probably have to run through the commit messages another, but if @lavalamp wants to take another look that would be nice.

I guess we can split this into two PRs as well, on for the `f.KnownObjects == nil` and one for `f.KnownObjects != nil`, in case you have preferences.

odinuge committed:
client-go/cache: rewrite Replace to check queue first

This is useful to both reduce the code complexity, and to ensure clients
get the "newest" version of an object known when its deleted. This is
all best-effort, but for clients it makes more sense giving them the
newest object they observed rather than an old one.

This is especially useful when an object is recreated. eg.

Object A with key K is in the KnownObjects store;
- DELETE delta for A is queued with key K
- CREATE delta for B is queued with key K
- Replace without any object with key K in it.

In this situation its better to create a DELETE delta with
DeletedFinalStateUnknown with B (with this patch), than it is to give
the client an DeletedFinalStateUnknown with A (without this patch).

Signed-off-by: Odin Ugedal <ougedal@palantir.com>
Signed-off-by: Odin Ugedal <odin@uged.al>

odinuge committed:
client-go/cache: update Replace comment to be more clear

Since the behavior is now changed, and the old behavior leaked objects,
this adds a new comment about how Replace works.

Signed-off-by: Odin Ugedal <ougedal@palantir.com>
Signed-off-by: Odin Ugedal <odin@uged.al>

aojea commented:
/assign @lavalamp 

this will need a release note

odinuge commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -652,6 +657,33 @@ func (f *DeltaFIFO) Replace(list []interface{}, _ string) error {
 	}
 
+	// detect deletions for items in the queue that are not yet present in knownObjects
+	for k, oldItem := range f.items {
```
Figured out we should probably do that at the same time as this to make the behavior consistent and to avoid more ambiguities. Pushed a new set of commits to do it.

odinuge commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -600,58 +599,56 @@ func (f *DeltaFIFO) Replace(list []interface{}, _ string) error {
+
 	if !f.populated {
 		f.populated = true
 		f.initialPopulationCount = keys.Len() + queuedDeletions
```
Maybe for another PR; but I'm wondering if doing this would make the logic simpler and safer? Especially around duplicate deletions etc.

```go
f.initialPopulationCount = len(f.queue)
```

lavalamp replied:
this only happens on the first replace, which always happens before any add/update/delete, so it can't hit the case you are fixing.

odinuge replied:
> ... I think. Not sure why it needs to worry about deletions.

Yup, same

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -600,58 +599,56 @@ func (f *DeltaFIFO) Replace(list []interface{}, _ string) error {
 		}
 	}
 
+	if f.knownObjects != nil {
```
I will read the comments in a second but I'm very confused about why you moved this block and switched its condition?

lavalamp replied:
ok -- read the comments and I'm still confused, can you put it back to minimize the diff or document why it has to change?

odinuge replied:
👋 Its mostly to get consistent behavior for when the object is present in `KnownObjects` or not when creating the deleted event with `DeletedFinalStateUnknown`.

Did you read the commit message? I'm happy to drop this and we can make a PR, or make it more clear via comments in the code.
https://github.com/kubernetes/kubernetes/pull/115620/commits/7bcc3e00fc28b2548886d04639a2e352ab37fb55


Also, this test change shows the actual "behavior change" https://github.com/kubernetes/kubernetes/pull/115620/commits/7bcc3e00fc28b2548886d04639a2e352ab37fb55#diff-b1aad36a628c7aeab74b3761643a965d90c258d019c4f77645ff45c51eb6ca35R498, as well as the rest of the test additions in `TestDeltaFIFO_ReplaceMakesDeletions`.

lavalamp replied:
ok, the test might have won me over, I will read this again in the morning. :)

lavalamp replied:
it makes sense to me now.

MikeSpreitzer commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -571,12 +571,11 @@ func (f *DeltaFIFO) Pop(process PopProcessFunc) (interface{}, error) {
+// object of K. The pre-existing keys are those in the union set of the keys in
+// `f.items` and `f.knownObjects` (if not nil). The last known object for key K is
+// the one present in the last delta in `f.items`. If there is no delta for K,
+// it is the object in `f.items`.
```
Is there a typo here, did you mean `f.knownObjects` rather than `f.items` in this case?

odinuge replied:
yes, correct, thanks!

MikeSpreitzer replied:
I do not see the correction here.

odinuge replied:
Was waiting on pushing in case we wanted more changes. I pushed a version with https://github.com/kubernetes/kubernetes/pull/115620#discussion_r1104863959 now 😄

MikeSpreitzer replied:
resolved

MikeSpreitzer commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -571,12 +571,11 @@ func (f *DeltaFIFO) Pop(process PopProcessFunc) (interface{}, error) {
+// `Delete(DeletedFinalStateUnknown{K, O})` where O is the latest known
+// object of K. The pre-existing keys are those in the union set of the keys in
+// `f.items` and `f.knownObjects` (if not nil). The last known object for key K is
+// the one present in the last delta in `f.items`. If there is no delta for K,
```
I presume that by "If there is no delta for K" you mean "If there is no Delta for K in `f.items`", right?

odinuge replied:
Yes.. Thanks!

Sooo;

```
// Replace atomically does two things: (1) it adds the given objects
// using the Sync or Replace DeltaType and then (2) it does some deletions.
// In particular: for every pre-existing key K that is not the key of
// an object in `list` there is the effect of
// `Delete(DeletedFinalStateUnknown{K, O})` where O is the latest known
// object of K. The pre-existing keys are those in the union set of the keys in
// `f.items` and `f.knownObjects` (if not nil). The last known object for key K is
// the one present in the last delta in `f.items`. If there is no delta for K
// in `f.items`, it is the object in `f.knownObjects`
```
looks good?

MikeSpreitzer replied:
yes, that would be better.

MikeSpreitzer replied:
resolved

MikeSpreitzer commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo_test.go:
```diff
@@ -121,6 +121,130 @@ func TestDeltaFIFO_replaceWithDeleteDeltaIn(t *testing.T) {
+			},
+		},
+		{
+			name: "Replaced object should have only a single Delete",
```
Maybe a better name here, differentiated from the following case?

odinuge replied:
yee, that sounds smart. I'll try to be a bit more creative and update the test names and push a new version tomorrow morning 😄 Thanks!

aojea commented:
> /lgtm /approve
> 
> Thanks!!

@lavalamp don't should be good to backport this change?

lavalamp commented:
Yeah I would definitely support a backport of this.

odinuge commented:
Sounds good! I'm happy to create backport PRs for this tomorrow or Monday. Been a while since a ran that script to generate cherry-picks :)

wojtek-t commented:
Thanks a lot for fixing this.

This LGTM too. And yes - we definitely should backport it.
